packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,246 @@
+      requiredApproval = ethers.constants.MaxUint256
+    } else {
+      /*requiredApproval = ethers.BigNumber.from(
+        tokenPaymasterRequest.feeQuote.maxGasFee.toString()
+      ).mul(
+        ethers.BigNumber.from(10).pow(
+          ethers.BigNumber.from(tokenPaymasterRequest.feeQuote.decimal.toString())
+        )
+      )*/
+
+      requiredApproval = BigNumber.from(
+        Math.floor(
+          tokenPaymasterRequest.feeQuote.maxGasFee *
+            Math.pow(10, tokenPaymasterRequest.feeQuote.decimal)
+        )
+      )
+    }
+
+    Logger.log('required approval for erc20 token ', requiredApproval)
+
+    const erc20 = new ethers.Contract(feeTokenAddress, ERC20_ABI, provider)
+
+    // TODO?
+    // Note: For some tokens we may need to set allowance to 0 first so that would return batch of transactions and changes the return type to Transaction[]
+    // In that case we would return two objects in an array, first of them being..
+    /*
+    {
+      to: erc20.address,
+      value: ethers.BigNumber.from(0),
+      data: erc20.interface.encodeFunctionData('approve', [spender, BigNumber.from("0")])
+    }
+    */
+
+    return {
+      to: erc20.address,
+      value: ethers.BigNumber.from(0),
+      data: erc20.interface.encodeFunctionData('approve', [spender, requiredApproval])
+    }
+  }
+
   /**
    *
    * @param userOp
-   * @description This function will send userOp to paymaster service to get signed from paymaster signer
-   * @returns string
+   * @param paymasterServiceData
+   * @returns
    */
-  async getPaymasterAndData(userOp: Partial<UserOperation>): Promise<string> {
+  async getPaymasterFeeQuotesOrData(
+    userOp: Partial<UserOperation>,
+    paymasterServiceData: FeeQuotesOrDataDto
+  ): Promise<BiconomyTokenPaymasterFeeQuoteResponse | string> {
+    userOp = await resolveProperties(userOp)
+    userOp.nonce = BigNumber.from(userOp.nonce).toHexString()
+    userOp.callGasLimit = BigNumber.from(userOp.callGasLimit).toHexString()
+    userOp.verificationGasLimit = BigNumber.from(userOp.verificationGasLimit).toHexString()
+    userOp.maxFeePerGas = BigNumber.from(userOp.maxFeePerGas).toHexString()
+    userOp.maxPriorityFeePerGas = BigNumber.from(userOp.maxPriorityFeePerGas).toHexString()
+    userOp.preVerificationGas = BigNumber.from(userOp.preVerificationGas).toHexString()
+    userOp.signature = '0x'
+    userOp.paymasterAndData = '0x'
+
+    let mode = null
+    let preferredToken = null
+    let feeTokensArray: string[] = []
+    // could make below null
+    let smartAccountInfo = {
+      name: 'BICONOMY',
+      version: '1.0.0'
+    }
+    let webhookData = null
+
+    if (paymasterServiceData.mode) {
+      Logger.log('Requested mode is ', paymasterServiceData.mode)
+      mode = paymasterServiceData.mode
+      // Validation on the mode passed / define allowed enums
+    }
+
+    if (paymasterServiceData.tokenInfo) {
+      if (paymasterServiceData.tokenInfo.preferredToken) {
```

Comment:
Two if can be combined into one using ? operator

Reply:
done